subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -175,25 +179,33 @@ public void beforeBuildFinished(boolean watchingEnabled) {
      */
     private synchronized void startWatching() {
         if (watchRegistry != null) {
+            if (!eventConsumerThread.isAlive()) {
+                throw new RuntimeException("The thread consuming the file events stopped for an unknown reason");
+            }
             return;
         }
         try {
             long startTime = System.currentTimeMillis();
             watchRegistry = watcherRegistryFactory.startWatcher(new FileWatcherRegistry.ChangeHandler() {
                 @Override
                 public void handleChange(FileWatcherRegistry.Type type, Path path) {
-                    try {
-                        fileEvents.put(new FileEvent(path, type));
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new RuntimeException(e);
+                    boolean addedToQueue = fileEvents.offer(FileEvent.changed(path, type));
+                    if (!addedToQueue) {
+                        LOGGER.warn("Gradle file event buffer overflow, dropping state");
+                        signalLostState();
                     }
                 }
 
                 @Override
                 public void handleLostState() {
+                    LOGGER.warn("Native file events buffer overflow, dropping state");
```

Comment:
This is not necessarily a native buffer overflow, can be an unknown event for example (logged elsewhere). I think we should just mention we lost state due to some native-side thing.

Reply:
I adjusted the log message.